modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb

```diff
@@ -0,0 +1,115 @@
+      classes = (current['classes'] || []).map do |clazz|
+        parameters = clazz['params'].map do |parameter|
+          if parameter['default_source'].is_a?(String) && parameter['default_source'].start_with?('$')
+            to_return = parameter.dup
+            to_return['default_source'] = "${#{parameter['default_source'].slice(1..-1)}}"
+            to_return
+          else
+            parameter
+          end
+        end
+        {'name' => clazz['name'], 'params' => parameters}
+      end
+      current.has_key?('error') ? current : {'path' => current['path'], 'classes' => classes}
+    end
+  end
+
+  def legacy_classes_format(response)
+    response['files'].map do |current|
+      (current['classes'] || []).map do |manifest| # current['classes'] is nil for 'error' entities
+        parameters = manifest['params'].inject({}) do |all, parameter|
+          parameter_value = parameter['default_literal'] || parameter['default_source']
+          all[parameter['name']] = (parameter_value.is_a?(String) && parameter_value.start_with?('$')) ? "${#{parameter_value.slice(1..-1)}}" : parameter_value
+          all
+        end
+        ::Proxy::Puppet::PuppetClass.new(manifest['name'], parameters)
+      end
+    end.flatten
+  end
+
+  def get_classes(environment)
+    future = async_get_classes(environment)
+    future.value!(MAX_CLIENT_TIMEOUT)
+
+    raise ::Proxy::Puppet::TimeoutError, "Puppet is taking too long to respond, please try again later." if future.pending?
+    future.value
+  end
+
+  def async_get_classes(environment)
+    etag, classes, existing_future = @m.synchronize { [@etag_cache[environment], @classes_cache[environment], @futures_cache[environment]] }
+    return existing_future unless existing_future.nil?
+
+    future = Concurrent::Promise.new do
+      begin
+        response, etag = @puppet_api.new(puppet_url, ssl_ca, ssl_cert, ssl_key).list_classes(environment, etag, MAX_PUPPETAPI_TIMEOUT)
+      rescue Exception => e
+        logger.error("Error while retrieving puppet classes for '%s' environment: %s" % [environment, e.backtrace])
+        raise e
+      end
+
+      if response == Proxy::PuppetApi::EnvironmentClassesApiv3::NOT_MODIFIED
+        @m.synchronize { @futures_cache[environment] = nil }
+        classes
+      else
+        @m.synchronize do
+          @futures_cache[environment] = nil
+          @etag_cache[environment] = etag
+          @classes_cache[environment] = response
+        end
+      end
+    end
+    @m.synchronize { @futures_cache[environment] = future }
+
+    future.execute
+  end
+end
+
+class Proxy::PuppetApi::EnvironmentClassesCacheInitializer
+  include ::Proxy::Log
+
+  def initialize(classes_retriever, environments_retriever)
+    @environments = environments_retriever
+    @classes = classes_retriever
+  end
+
+  def start
+    logger.info("Started puppet class cache initialization")
+    futures = @environments.all.map(&:name).map do |environment|
```

Comment:
The environment retrieval here is in the main service startup thread, so when the Puppet Server's inaccessible or slow, the module fails to load. Can you move it to a thread?

<pre>
E, [2016-09-23T11:58:55.600886 #2886] ERROR -- : Couldn't enable 'puppet_proxy_puppet_api': Failed to open TCP connection to foreman.example.com:8140 (Connection refused - connect(2) for "foreman.example.com" port 8140)
D, [2016-09-23T11:58:55.600921 #2886] DEBUG -- : ["/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:882:in `rescue in block in connect'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:879:in `block in connect'", "/home/dcleal/.rvm/rubies/ru
by-2.3.0/lib/ruby/2.3.0/timeout.rb:91:in `block in timeout'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/timeout.rb:101:in `timeout'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:878:in `connect'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby
/2.3.0/net/http.rb:863:in `do_start'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:852:in `start'", "/home/dcleal/.rvm/rubies/ruby-2.3.0/lib/ruby/2.3.0/net/http.rb:1398:in `request'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_common/a
pi_request.rb:39:in `send_request'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_puppet_api/v3_api_request.rb:4:in `find_environments'", "/home/dcleal/code/foreman/smart-proxy/modules/puppet_proxy_common/environments_retriever_base.rb:18:in `all'", "/home/
dcleal/code/foreman/smart-proxy/modules/puppet_proxy_puppet_api/v3_environment_classes_api_classes_retriever.rb:109:in `start'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:228:in `block in start_services'", "/home/dcleal/code/foreman/smart-prox
y/lib/proxy/plugin_initializer.rb:226:in `each'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:226:in `start_services'", "/home/dcleal/code/foreman/smart-proxy/lib/proxy/plugin_initializer.rb:211:in `configure_plugin'", "/home/dcleal/code/foreman
</pre>

(This is very likely at server startup as PS is much slower than foreman-proxy to load.)